### PR TITLE
ENH: Wrap string ufuncs in np.strings to allow default arguments

### LINKED ISSUE #25566: ENH: Wrap string ufuncs in np.strings to allow optional / default arguments
Strings ufuncs accept arguments to configure various things, like maximum number of replacements in `replace` or start and end position of possible matches in `find`. The CPython API provides sensible defaults for those parameters, so that uses don't have to do things like `'ttt'.replace('t', 'a', 3)`. We should do the same in `np.strings` and wrap the ufuncs to allow optional keyword arguments with defaults.

That way the naming of the parameters will be improved as well, so that we have `endswith(buf, suffix, start, end)` instead of `endswith(x1, x2, x3, x4)`.

Proposed in https://github.com/numpy/numpy/pull/25558#issuecomment-1885187601.

### PR DESCRIPTION
**Pull request by lysnikolaou**
Closes #25566.

### CODE REVIEW

File: numpy/_core/strings.py
Comment (mhvk) [inline numpy/_core/strings.py:130]: I'd put this in for `find` and `rfind` too.

File: numpy/_core/strings.py
Comment (mhvk) [inline numpy/_core/strings.py:132]: See comment for `find`

File: numpy/_core/strings.py
Comment (mhvk) [inline numpy/_core/strings.py:165]: See comment for `find`

File: numpy/_core/strings.pyi
Comment (mhvk) [inline numpy/_core/strings.pyi:56]: I think you need to change the argument names here too...

**Review by mhvk — COMMENTED** [review]
I think this is for the better but as noted inline, we mostly don't need to do `asanyarray` (`replace` was an exception since otherwise one had to cast to array many times). The advantage is that the functions can be overridden with `__array_ufunc__`.

A question is whether we should also allow overrides with `__array_function__` and thus provide a dispatcher. In general, I think every non-ufunc is meant to be overrideable with that, though if we do not cast to array at all, then arguably it makes little sense to do so. ping @ahaldane.

p.s. I'm wondering if there would be a point for `ufunc`s to be able to provide defaults... But probably best to make those not yet more complicated.

**Comment by lysnikolaou** [comment]
I'm not knowledgeable enough to have an opinion on the `__array_function__` discussion. Either way, can we please try to get this in as soon as possible, cause it's blocking all of my work on writing more `np.strings` ufuncs?

File: numpy/_core/strings.py
Comment (mhvk) [inline numpy/_core/strings.py:72]: Sorry to be nitpicky, but I'd remove `, 0, None` from the example, since those are defaults.

**Review by mhvk — APPROVED** [review]
One total nitpick left.

The typing changes I cannot judge as well, but seem reasonable.

Will approve now, but not yet merge so you can address my nitpick (and @ngoldbaum can have a last look).

**Comment by ngoldbaum** [comment]
I'll go ahead and push a fixup for this so @lysnikolaou is unblocked during European working hours tomorrow. Otherwise LGTM.

btw @lysnikolaou assuming no one objects to NEP 55 next week and stringdtype is merged within the next week or two, would you be OK with holding off on merging more ufuncs until stringdtype is merged? I'm happy to do this how we've been doing it but instead we just make sure all three dtypes get implementations before merging the PR - you implement it first for bytes and unicode and then I push fixups to the PR for stringdtype so all three are supported when the PR is merged.

The other way we could do it is make sure that all the functions (including the ones we've already converted) fall back to `_vec_string` if the dtypes aren't recognized as explicitly supported numpy dtypes (e.g. `dtype.char not in 'SUT'`). That way user string dtypes will continue to work (albeit slowly) without needing to implement a ufunc, as long as their scalar type is a string subtype. Asking user dtypes to implement ufuncs is fine, but it was also quite nice for prototyping stringdtype that I didn't need to implement a bunch of ufuncs to get the `np.char` functions working.

**Comment by lysnikolaou** [comment]
> you implement it first for bytes and unicode and then I push fixups to your PR for stringdtype so all three are supported when the PR is merged

This seems doable. My plan is to merge this PR, open a new one that uses `np.strings` in `np.char` and then slowly start to open the ufuncs PRs. I have already implemented lots of them, so I'll probably open them in batches.

**Pull request merged**
